File: apps/android/core/model/src/main/kotlin/me/meeshy/sdk/model/call/CallRetryPolicy.kt
package me.meeshy.sdk.model.call

/**
 * Pure decision: which [CallEndReason]s warrant a « Réessayer » (retry)
 * affordance. Only TRANSIENT establishment/drop failures — a retry genuinely
 * recovers those (prod 2026-07-12: ~16% of calls end in Failed/ConnectionLost,
 * commonly transient ICE-gathering / TURN-allocation hiccups). Normal outcomes
 * (Local/Remote hangup, Missed, Rejected) are never retried.
 *
 * Parité web `isRetryableCallFailure` (apps/web/lib/calls/call-retry-policy.ts)
 * and the reliability aggregate's failure set — one rule, three platforms.
 */
object CallRetryPolicy {
    fun isRetryable(reason: CallEndReason?): Boolean = when (reason) {
        is CallEndReason.Failed, CallEndReason.ConnectionLost -> true
        else -> false
    }
}

File: apps/web/lib/calls/call-retry-policy.ts
/**
 * Pure decision: which call end reasons warrant a « Réessayer » (retry)
 * affordance. Only TRANSIENT establishment/drop failures — a retry genuinely
 * recovers those (prod 2026-07-12: ~16% of calls end in failed/connectionLost,
 * commonly transient ICE-gathering / TURN-allocation hiccups). Normal outcomes
 * (completed/missed/rejected) and server-side death (heartbeatTimeout/
 * garbageCollected) are never retried — a retry there would be surprising
 * (declined) or futile (the peer's side is gone).
 */

import type { CallEndReason } from '@meeshy/shared/types/video-call';

const RETRYABLE: ReadonlySet<CallEndReason> = new Set<CallEndReason>([
  'failed',
  'connectionLost',
]);

export function isRetryableCallFailure(reason: CallEndReason | null | undefined): boolean {
  return reason != null && RETRYABLE.has(reason);
}
